network/src/peers/peers.rs

```diff
@@ -344,4 +387,18 @@ impl Node {
             self.peer_book.write().add_peer(peer_address);
         }
     }
+
+    fn can_connect(&self) -> bool {
+        let peer_book = self.peer_book.read();
```

Comment:
it's extremely unlikely to matter, but just due to the fact that there's a potential call to `warn!` which, as all `stdout` operations, is quite slow, I'd limit the scope of this lock by introducing a "synthetic" one that returns a pair with `num_connected` and `num_connecting`.